### PR TITLE
[WASMFS] getdents syscall

### PR DESCRIPTION
**Pull request by ethanalee**
Relevant Issue: #15041 

- Implement `getdents` syscall.
- Add relevant tests.

### CODE REVIEW

File: system/lib/wasmfs/syscalls.cpp [hunk 1]
```diff
@@ -513,4 +515,67 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+
+  // Check if the result buffer is too small.
+  if (count / sizeof(dirent) == 0) {
+    return -EINVAL;
+  }
+
+  auto file = openFile.locked().getFile();
+
+  auto directory = file->dynCast<Directory>();
+  // Check if the file is a directory.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:535]: In general, I think the previous comment is good - it's not obvious when reading the `count / sizeoff` stuff what the meaning is.

But this type of comment seems redundant: the code literally does `if (!directory)`, so I would remove it. Just phrasing the code in human language doesn't add much, I think, unless it also adds context or information.

File: system/lib/wasmfs/syscalls.cpp [hunk 2]
```diff
@@ -513,4 +515,67 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+  off_t bytesRead = 0;
+  // A directory's position corresponds to the index in its entries vector.
+  int index = openFile.locked().position();
+
+  // In the root directory, ".." refers to itself.
+  auto dotdot =
+    file == wasmFS.getRootDirectory() ? nullptr : lockedDir.getParent();
+
+  // There are always two hardcoded directories "." and ".."
+
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:552]:
```suggestion
```

File: system/lib/wasmfs/syscalls.cpp [hunk 3]
```diff
@@ -513,4 +515,67 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+#ifdef WASMFS_DEBUG
+  for (auto pair : entries) {
+    emscripten_console_log(pair.name.c_str());
+  }
+#endif
+
+  for (; index < entries.size() && bytesRead + sizeof(dirent) <= count;
+       index++) {
+    auto curr = entries[index];
+    result->d_ino = (ino_t)curr.file.get(); // Set inode number to file pointer.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:566]:
```suggestion
    // Set inode number to file pointer. This gives a unique inode number for
    // each file in the simplest possible way.
    // TODO: For security it would probably be better to use an indirect mapping.
    result->d_ino = (ino_t)curr.file.get();
```
I believe we discussed that in the meeting?

File: tests/wasmfs/wasmfs_getdents.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  print(d, fd);
+
+  // Try reading an invalid fd.
+  errno = 0;
+  getdents(-1, &d, sizeof(d));
+  printf("Errno: %s\n", strerror(errno));
+  assert(errno == EBADF);
+
+  // Try passing in a size that is too small.
+  // Existing file system doesn't catch this error.
```
Comment (kripken) [inline tests/wasmfs/wasmfs_getdents.c:52]: This comment is dangerous as it will get stale once WasmFS is the default FS.

How about "the JS filesystem"?

File: system/lib/wasmfs/syscalls.cpp [hunk 5]
```diff
@@ -513,4 +537,69 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+  }
+#endif
+
+  for (; index < entries.size() && bytesRead + sizeof(dirent) <= count;
+       index++) {
+    auto curr = entries[index];
+    // Set inode number to the file pointer. This gives a unique inode number
+    // for each file in the simplest possible way.
+    // TODO: For security it would be better to use an indirect mapping.
+    result->d_ino = (ino_t)curr.file.get();
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:590]: Looks like this pattern appears in multiple PRs. How about adding a helper function "get ino", and that would be the single place we do this conversion? Then we can put the security comment just there, and also a static assert as I mentioned in the other PR on the size of `ino_t` being big enough for a pointer.

File: system/lib/wasmfs/syscalls.cpp [hunk 6]
```diff
@@ -386,6 +405,11 @@ long __syscall_mkdir(long path, long mode) {
   if (curr) {
     return -EEXIST;
   } else {
+    // Mask rwx permissions for user, group and others, and the sticky bit.
+    // This prevents users from entering S_IFREG for example.
+    // https://www.gnu.org/software/libc/manual/html_node/Permission-Bits.html
+    // This is a placeholder for now which matches the JS filesystem.
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:411]: We determined that this was important to prevent users from setting bits that they shouldn't have control over.
```suggestion
```

File: system/lib/wasmfs/syscalls.cpp [hunk 7]
```diff
@@ -327,6 +336,16 @@ __wasi_fd_t __syscall_open(long pathname, long flags, long mode) {
...
     if (flags & O_CREAT) {
+      // Since mode is optional, mode is specified using varargs.
+      mode_t mode = 0;
+      va_list vl;
+      va_start(vl, flags);
+      mode = va_arg(vl, int);
+      va_end(vl);
+      // Mask all permissions sent via mode.
+      // This is a placeholder for now which matches the JS filesystem.
+      mode &= S_IALLUGO;
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:347]: Why use `S_IALLUGO` here and `S_IRWXUGO` below? What's the difference? (As we discussed in chat, it would be nice to have our own more readable macros for permissions.)

File: system/lib/wasmfs/syscalls.cpp [hunk 8]
```diff
@@ -513,4 +537,69 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+  // Hold the locked directory for the duration of this operation.
+  auto lockedDir = directory->locked();
+
+  off_t bytesRead = 0;
+  // A directory's position corresponds to the index in its entries vector.
+  int index = openFile.locked().position();
+
+  // In the root directory, ".." refers to itself.
+  auto dotdot =
+    file == wasmFS.getRootDirectory() ? nullptr : lockedDir.getParent();
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:571]: The `..` for the root directory should be itself. But if the directory we are reading is unlinked, it's still possible for `dotdot` to be null. I'm not sure whether we should include the `..` entry in that case, or if we should, what value it should have. Could you do some experiments to find out?

**Review by tlively — COMMENTED** [review]
It looks like this still includes changes from #15445, so it would be good to merge `main` in.

File: system/lib/wasmfs/syscalls.cpp (hunk 7 above)
Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:347]: (Ignore this comment for this PR if this is entirely a change from #15445)

Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:347]: Yep, these changes are from the previous PR.

Comment (tlively) [reply system/lib/wasmfs/syscalls.cpp:346]: Can you do a merge (or if you have to, a rebase) to eliminate the unrelated changes from the diff on this PR?

Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:346]: Is that possible since I merged the previous PR into this branch instead of main?

File: system/lib/wasmfs/syscalls.cpp (hunk 8 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:570]: I think you can't unlink a directory. I tried it out with `rmdir`, but I you can't remove the upper directory (`..`) since it has children.

File: system/lib/wasmfs/syscalls.cpp (hunk 5 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp]: Done

File: system/lib/wasmfs/syscalls.cpp [hunk 9]
```diff
@@ -513,4 +536,73 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+  for (auto pair : entries) {
+    emscripten_console_log(pair.name.c_str());
+  }
+#endif
+
+  for (; index < entries.size() && bytesRead + sizeof(dirent) <= count;
+       index++) {
+    auto curr = entries[index];
+    // Set inode number to the file pointer. This gives a unique inode number
+    // for each file in the simplest possible way.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:592]:
```suggestion
```
This comment is now stale. It is not needed here, and it just adds risk of becoming incorrect some day if we change the impl in the other file.

Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp]: Moved it to the `getIno` function. I think it's relevant there.

File: system/lib/wasmfs/syscalls.cpp [hunk 10]
```diff
@@ -535,4 +535,72 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+
+  for (; index < entries.size() && bytesRead + sizeof(dirent) <= count;
+       index++) {
+    auto curr = entries[index];
+    result->d_ino = curr.file->locked().getIno();
+    result->d_off = bytesRead + sizeof(dirent);
+    result->d_reclen = sizeof(dirent);
+    result->d_type =
+      curr.file->is<Directory>() ? DT_DIR : DT_REG; // TODO: add symlinks.
+    strcpy(result->d_name, curr.name.c_str());
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:596]: Can we add an assertion that the name fits into the `d_name` field? We also need to enforce that somehow, which I don't think we currently do, although enforcement and testing can be in a separate PR.

File: system/lib/wasmfs/file.h [hunk 11]
```diff
@@ -63,6 +63,14 @@ class File : public std::enable_shared_from_this<File> {
     time_t& mtime() { return file->mtime; }
     time_t& atime() { return file->atime; }
 
+    ino_t getIno() {
+      // Set inode number to the file pointer. This gives a unique inode number.
+      // TODO: For security it would be better to use an indirect mapping.
+      // Ensure that the pointer will not overflow an ino_t.
+      static_assert(sizeof(file.get()) <= sizeof(ino_t));
+      return (ino_t)file.get();
+    }
```
Comment (tlively) [inline system/lib/wasmfs/file.h:72]: This can probably be a method directly on `File` rather than `File::Handle`, since the inode and underlying address are immutable. Taking the lock does not have any benefit here.

**Review by tlively — APPROVED** [review]
LGTM with those final comments resolved 👍

File: system/lib/wasmfs/file.h (hunk 11 above)
Comment (ethanalee) [reply system/lib/wasmfs/file.h:72]: I see, locking doesn't have any benefit here it seems.

File: system/lib/wasmfs/file.h [hunk 12]
```diff
@@ -169,11 +183,19 @@ class Directory : public File {
       return "";
     }
 
+    // Return a vector of the key-value pairs in entries.
+    auto getEntries() {
```
[the remark this answers is not in the document]
Comment (ethanalee) [reply system/lib/wasmfs/file.h:187]: Sounds good, I originally did it since the return type was quite verbose previously. I think with the struct return type it is a moot point. 👍

File: system/lib/wasmfs/syscalls.cpp (hunk 10 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:596]: Done

Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:596]: Done

**Review by kripken — APPROVED** [review]

**Comment by tlively** [comment]
@ethanalee, these test failures look like they might be real. Could you take a look?
